From 3b0c8ceb6471845730ae9c64defb8dbdf74fac18 Mon Sep 17 00:00:00 2001
From: Tommy Webb <tommy@calyxinstitute.org>
Date: Tue, 11 Jun 2024 18:52:16 +0000
Subject: [PATCH 1/3] Multiple work profile badge colors for Launcher3

Original-Change-Id: I64f1b425eee5cc7994616050411a881d84ece99d
Original-Change-Id-2: I9b9f01a8e483ee3de1c11a82e8b569537308596d
Change-Id: Id1541fe766dfe12a458e42fb2c26fcb51b9003a6
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 iconloaderlib/res/values-night-v31/colors.xml |  2 +
 iconloaderlib/res/values-night/colors.xml     |  2 +
 iconloaderlib/res/values-v31/colors.xml       |  2 +
 iconloaderlib/res/values/colors.xml           |  7 +-
 .../launcher3/icons/BaseIconFactory.java      | 86 +++++++++++++++++++
 .../com/android/launcher3/icons/BitmapInfo.kt | 27 ++++++
 .../launcher3/icons/cache/BaseIconCache.kt    | 12 ++-
 7 files changed, 131 insertions(+), 7 deletions(-)

diff --git a/iconloaderlib/res/values-night-v31/colors.xml b/iconloaderlib/res/values-night-v31/colors.xml
index e5ebda6..5ed9e2a 100644
--- a/iconloaderlib/res/values-night-v31/colors.xml
+++ b/iconloaderlib/res/values-night-v31/colors.xml
@@ -19,6 +19,8 @@
 <resources>
     <color name="themed_icon_color">@android:color/system_accent1_200</color>
     <color name="themed_icon_background_color">@android:color/system_accent2_800</color>
+    <!-- Fallback colors specified in main colors.xml
     <color name="themed_badge_icon_color">@android:color/system_accent2_800</color>
     <color name="themed_badge_icon_background_color">@android:color/system_accent1_200</color>
+    -->
 </resources>
diff --git a/iconloaderlib/res/values-night/colors.xml b/iconloaderlib/res/values-night/colors.xml
index 9de7074..f7b1558 100644
--- a/iconloaderlib/res/values-night/colors.xml
+++ b/iconloaderlib/res/values-night/colors.xml
@@ -19,6 +19,8 @@
 <resources>
     <color name="themed_icon_color">#A8C7FA</color>
     <color name="themed_icon_background_color">#003355</color>
+    <!-- Fallback colors specified in main colors.xml
     <color name="themed_badge_icon_color">#003355</color>
     <color name="themed_badge_icon_background_color">#A8C7FA</color>
+    -->
 </resources>
diff --git a/iconloaderlib/res/values-v31/colors.xml b/iconloaderlib/res/values-v31/colors.xml
index 1405ad0..b1e7492 100644
--- a/iconloaderlib/res/values-v31/colors.xml
+++ b/iconloaderlib/res/values-v31/colors.xml
@@ -19,6 +19,8 @@
 <resources>
     <color name="themed_icon_color">@android:color/system_accent1_700</color>
     <color name="themed_icon_background_color">@android:color/system_accent1_100</color>
+    <!-- Fallback colors specified in main colors.xml
     <color name="themed_badge_icon_color">@android:color/system_accent1_700</color>
     <color name="themed_badge_icon_background_color">@android:color/system_accent1_100</color>
+    -->
 </resources>
diff --git a/iconloaderlib/res/values/colors.xml b/iconloaderlib/res/values/colors.xml
index 56ae0b6..6b7a484 100644
--- a/iconloaderlib/res/values/colors.xml
+++ b/iconloaderlib/res/values/colors.xml
@@ -19,11 +19,12 @@
 <resources>
     <color name="themed_icon_color">#0842A0</color>
     <color name="themed_icon_background_color">#D3E3FD</color>
-    <color name="themed_badge_icon_color">#0842A0</color>
-    <color name="themed_badge_icon_background_color">#D3E3FD</color>
+    <!-- Grayscale work badge icon color used for fallback only (system provides colorized icon) -->
+    <color name="themed_badge_icon_color">#AAAAAA</color>
+    <color name="themed_badge_icon_background_color">#222222</color>
 
     <color name="badge_tint_instant">@android:color/black</color>
-    <color name="badge_tint_work">#1A73E8</color>
+    <color name="badge_tint_work">@color/themed_badge_icon_color</color>
     <color name="badge_tint_private">#3C4043</color>
     <color name="badge_tint_clone">#ff3C4043</color>
 
diff --git a/iconloaderlib/src/com/android/launcher3/icons/BaseIconFactory.java b/iconloaderlib/src/com/android/launcher3/icons/BaseIconFactory.java
index 1f107c8..5b33773 100644
--- a/iconloaderlib/src/com/android/launcher3/icons/BaseIconFactory.java
+++ b/iconloaderlib/src/com/android/launcher3/icons/BaseIconFactory.java
@@ -34,7 +34,9 @@ import android.graphics.drawable.ColorDrawable;
 import android.graphics.drawable.Drawable;
 import android.graphics.drawable.InsetDrawable;
 import android.os.Build;
+import android.os.Process;
 import android.os.UserHandle;
+import android.util.Pair;
 import android.util.SparseArray;
 
 import androidx.annotation.ColorInt;
@@ -48,6 +50,8 @@ import com.android.launcher3.util.FlagOp;
 import com.android.launcher3.util.UserIconInfo;
 
 import java.lang.annotation.Retention;
+import java.util.LinkedHashMap;
+import java.util.Map;
 
 /**
  * This class will be moved to androidx library. There shouldn't be any dependency outside
@@ -101,6 +105,19 @@ public class BaseIconFactory implements AutoCloseable {
 
     private int mWrapperBackgroundColor = DEFAULT_WRAPPER_BACKGROUND;
 
+    // User badges cached by size, e.g. workspace badge (large) vs widget badge (small)
+    private final LinkedHashMap<Pair<UserHandle, Integer>, Bitmap> mUserBadges =
+            new LinkedHashMap<>() {
+        // This *is* a cache, so it shouldn't grow forever. Lazily limit it by number of entries.
+        // We will likely only be dealing with 2 different sizes; multiply that by the number of
+        // profiles + 1 for the current user, and we really don't need much room. This is plenty.
+        private static final int MAX_ENTRIES = 50;
+        @Override
+        protected boolean removeEldestEntry(Map.Entry eldest) {
+            return size() > MAX_ENTRIES;
+        }
+    };
+
     private static int PLACEHOLDER_BACKGROUND_COLOR = Color.rgb(245, 245, 245);
 
     protected BaseIconFactory(Context context, int fullResIconDpi, int iconBitmapSize,
@@ -240,6 +257,13 @@ public class BaseIconFactory implements AutoCloseable {
                     )
             );
         }
+        if (options != null) {
+            final UserHandle user = options.mUserHandle != null ? options.mUserHandle
+                    : (options.mUserIconInfo != null ? options.mUserIconInfo.user : null);
+            if (user != null) {
+                info.setUser(user, this);
+            }
+        }
         FlagOp flagOp = getBitmapFlagOp(options);
         if (adaptiveIcon instanceof WrappedAdaptiveIcon) {
             flagOp = flagOp.addFlag(BitmapInfo.FLAG_WRAPPED_NON_ADAPTIVE);
@@ -536,6 +560,68 @@ public class BaseIconFactory implements AutoCloseable {
         return createBadgedIconBitmap(iconProvider.getFullResDefaultActivityIcon(mFullResIconDpi));
     }
 
+    public Drawable getBadgeForUser(UserHandle user) {
+        return getBadgeForUser(user, mIconBitmapSize);
+    }
+
+    /**
+     * Returns a drawable that can be used as a badge for the user or null.
+     */
+    // @UiThread
+    public Drawable getBadgeForUser(UserHandle user, int iconSize) {
+        final int badgeSize = getBadgeSizeForIconSize(iconSize);
+        Bitmap badgeBitmap = getUserBadge(user, badgeSize);
+        FastBitmapDrawable d = new FastBitmapDrawable(badgeBitmap);
+        d.setFilterBitmap(true);
+        d.setBounds(0 /* left */, 0 /* top */, badgeBitmap.getWidth(), badgeBitmap.getHeight());
+        return d;
+    }
+
+    private Bitmap getUserBadge(UserHandle user, int badgeSize) {
+        if (badgeSize > mIconBitmapSize) {
+            throw new IllegalArgumentException("badgeSize cannot be larger than mIconBitmapSize: "
+                    + "got " + badgeSize + ", expected less than or equal to " + mIconBitmapSize);
+        }
+        Pair<UserHandle, Integer> userBadgeOfSize = new Pair<>(user, badgeSize);
+        synchronized (mUserBadges) {
+            Bitmap badgeBitmap = mUserBadges.get(userBadgeOfSize);
+            if (badgeBitmap != null) {
+                return badgeBitmap;
+            }
+
+            final Resources res = mContext.getResources();
+            Bitmap badgedBitmap = Bitmap.createBitmap(
+                    mIconBitmapSize, mIconBitmapSize, Bitmap.Config.ARGB_8888);
+
+            // PackageManager's getUserBadgedDrawableForDensity results in a giant work profile
+            // icon that extends outside of the badge icon's circle, seemingly no matter what
+            // arguments are provided. getUserBadgeForDensity is not even exposed (hidden).
+            // So, unfortunately, we draw into a full icon size Bitmap and then crop out the
+            // badge from the corner.
+            Drawable drawable = mContext.getPackageManager().getUserBadgedIcon(
+                    new BitmapDrawable(res, badgedBitmap), user);
+            /* Drawable drawable = mContext.getPackageManager().getUserBadgedDrawableForDensity(
+                    new BitmapDrawable(res, badgeBitmap), user,
+                    new Rect(0, 0, badgeSize, badgeSize),
+                    0); */
+            if (drawable instanceof BitmapDrawable) {
+                badgedBitmap = ((BitmapDrawable) drawable).getBitmap();
+            } else {
+                badgedBitmap.eraseColor(Color.TRANSPARENT);
+                Canvas c = new Canvas(badgedBitmap);
+                drawable.setBounds(0, 0, badgeSize, badgeSize);
+                drawable.draw(c);
+                c.setBitmap(null);
+            }
+            final int cropOffset = Math.max(mIconBitmapSize - badgeSize, 0);
+            badgeBitmap = Bitmap.createBitmap(badgedBitmap,
+                    cropOffset /* x */, cropOffset /* y */,
+                    badgeSize /* width */, badgeSize /* height */);
+            mUserBadges.put(userBadgeOfSize, badgeBitmap);
+            return badgeBitmap;
+        }
+    }
+
     /**
      * Returns the correct badge size given an icon size
      */
diff --git a/iconloaderlib/src/com/android/launcher3/icons/BitmapInfo.kt b/iconloaderlib/src/com/android/launcher3/icons/BitmapInfo.kt
index 1b3f0fa..c6b9986 100644
--- a/iconloaderlib/src/com/android/launcher3/icons/BitmapInfo.kt
+++ b/iconloaderlib/src/com/android/launcher3/icons/BitmapInfo.kt
@@ -20,6 +20,7 @@ import android.graphics.Bitmap
 import android.graphics.Canvas
 import android.graphics.Path
 import android.graphics.drawable.Drawable
+import android.os.UserHandle
 import androidx.annotation.ColorRes
 import androidx.annotation.DrawableRes
 import androidx.annotation.IntDef
@@ -46,6 +47,10 @@ open class BitmapInfo(
 
     private var badgeInfo: BitmapInfo? = null
 
+    private var userHandle: UserHandle? = null
+
+    private var userBadge: Drawable? = null
+
     fun withBadgeInfo(badgeInfo: BitmapInfo?) = clone().also { it.badgeInfo = badgeInfo }
 
     /** Returns a bitmapInfo with the flagOP applied */
@@ -56,6 +61,10 @@ open class BitmapInfo(
         return clone().also { it.flags = op.apply(it.flags) }
     }
 
+    fun withUser(user: UserHandle?, iconFactory: BaseIconFactory): BitmapInfo {
+        return clone().also { it.setUser(user, iconFactory) }
+    }
+
     @Override
     open fun clone(): BitmapInfo {
         return copyInternalsTo(BitmapInfo(icon, color))
@@ -65,9 +74,20 @@ open class BitmapInfo(
         target.themedBitmap = themedBitmap
         target.flags = flags
         target.badgeInfo = badgeInfo
+        target.userHandle = userHandle
+        target.userBadge = userBadge
         return target
     }
 
+    fun setUser(user: UserHandle?, iconFactory: BaseIconFactory) {
+        userHandle = user
+        if (user != null) {
+            userBadge = iconFactory.getBadgeForUser(userHandle)
+        } else {
+            userBadge = null
+        }
+    }
+
     // TODO: rename or remove because icon can no longer be null?
     val isNullOrLowRes: Boolean
         get() = icon == LOW_RES_ICON
@@ -86,6 +106,8 @@ open class BitmapInfo(
         return !isNullOrLowRes
     }
 
+    fun getUser(): UserHandle? = userHandle
+
     /** Creates a drawable for the provided BitmapInfo */
     @JvmOverloads
     fun newIcon(
@@ -174,6 +196,11 @@ open class BitmapInfo(
         }
         if (skipUserBadge) {
             return null
+        } else if (userBadge != null) {
+            // We use a copy of the badge, or changes will affect everywhere it is used;
+            // e.g., shortcuts/widget user badges are very small, and these could affect
+            // regular launcher icons, and the other way around.
+            return userBadge!!.constantState?.newDrawable()?.mutate()
         } else {
             getBadgeDrawableInfo()?.let {
                 return UserBadgeDrawable(
diff --git a/iconloaderlib/src/com/android/launcher3/icons/cache/BaseIconCache.kt b/iconloaderlib/src/com/android/launcher3/icons/cache/BaseIconCache.kt
index eb8c5a2..9c6e3cb 100644
--- a/iconloaderlib/src/com/android/launcher3/icons/cache/BaseIconCache.kt
+++ b/iconloaderlib/src/com/android/launcher3/icons/cache/BaseIconCache.kt
@@ -242,10 +242,12 @@ constructor(
 
     @Synchronized
     fun getDefaultIcon(user: UserHandle): BitmapInfo {
-        if (defaultIcon == null) {
-            iconFactory.use { li -> defaultIcon = li.makeDefaultIcon(iconProvider) }
+        iconFactory.use { li ->
+            if (defaultIcon == null) {
+                defaultIcon = li.makeDefaultIcon(iconProvider)
+            }
+            return defaultIcon!!.withUser(user, li)
         }
-        return defaultIcon!!.withFlags(getUserFlagOpLocked(user))
     }
 
     protected fun getUserFlagOpLocked(user: UserHandle): FlagOp {
@@ -571,7 +573,9 @@ constructor(
             }
         }
         entry.bitmap.flags = c.getInt(INDEX_FLAGS)
-        entry.bitmap = entry.bitmap.withFlags(getUserFlagOpLocked(cacheKey.user))
+        iconFactory.use { factory ->
+            entry.bitmap = entry.bitmap.withUser(cacheKey.user, factory)
+        }
         iconProvider.notifyIconLoaded(entry.bitmap, cacheKey, logic)
         return true
     }
-- 
2.52.0

